internals/secrethub/service_gcp_init.go

```diff
@@ -0,0 +1,302 @@
+type ServiceGCPInitCommand struct {
+	description         string
+	repo                api.RepoPath
+	kmsKeyResourceID    string
+	serviceAccountEmail string
+	permission          string
+	io                  ui.IO
+	newClient           newClientFunc
+}
+
+// NewServiceGCPInitCommand creates a new ServiceGCPInitCommand.
+func NewServiceGCPInitCommand(io ui.IO, newClient newClientFunc) *ServiceGCPInitCommand {
+	return &ServiceGCPInitCommand{
+		io:        io,
+		newClient: newClient,
+	}
+}
+
+// Run initializes an GCP service.
+func (cmd *ServiceGCPInitCommand) Run() error {
+	client, err := cmd.newClient()
+	if err != nil {
+		return err
+	}
+
+	if cmd.serviceAccountEmail == "" && cmd.kmsKeyResourceID == "" {
+		fmt.Fprintln(cmd.io.Stdout(), "This command creates a new service account for use on GCP. For help on this, run `secrethub service gcp init --help`.")
+
+		var projectID string
+		creds, err := transport.Creds(context.Background())
+		if err == nil && creds.ProjectID != "" {
+			projectID = creds.ProjectID
+		} else {
+			var projectLister gcpProjectOptionLister
+			chosenProjectID, err := ui.ChooseDynamicOptions(cmd.io, "What GCP project do you want to use?", projectLister.Options, true, "project")
+			if err != nil {
+				return err
+			}
+
+			projectID = chosenProjectID
+		}
+
+		serviceAccountLister := gcpServiceAccountOptionLister{
+			ProjectID: projectID,
+		}
+		serviceAccountEmail, err := ui.ChooseDynamicOptionsValidate(cmd.io, "What is the email of the service account you want to use?", serviceAccountLister.Options, "service account", api.ValidateGCPServiceAccountEmail)
+		if err != nil {
+			return err
+		}
+		cmd.serviceAccountEmail = serviceAccountEmail
+
+		kmsKeyLister, err := newGCPKeyOptionsLister(projectID)
+		if err != nil {
+			return err
+		}
+		keyring, err := ui.ChooseDynamicOptions(cmd.io, "In which keyring is the KMS key you want to use for encrypting the service account's key?", kmsKeyLister.KeyringOptions, true, "keyring")
+		if err != nil {
+			return err
+		}
+		kmsKey, err := ui.ChooseDynamicOptions(cmd.io, "What is the KMS key you want to use for encrypting the service account's key?", kmsKeyLister.KeyOptions(keyring), true, "kms key")
+		if err != nil {
+			return err
+		}
+		cmd.kmsKeyResourceID = kmsKey
+	}
+
+	if cmd.serviceAccountEmail == "" {
+		serviceAccountEmail, err := ui.AskAndValidate(cmd.io, "What is the email of the GCP Service Account that should have access to the service?\n", 3, api.ValidateGCPServiceAccountEmail)
+		if err != nil {
+			return err
+		}
+		cmd.serviceAccountEmail = strings.TrimSpace(serviceAccountEmail)
+	}
+
+	if cmd.kmsKeyResourceID == "" {
+		kmsKey, err := ui.AskAndValidate(cmd.io, "What is the Resource ID of the KMS-key that should be used for encrypting the service's account key?\n", 3, api.ValidateGCPKMSKeyResourceID)
+		if err != nil {
+			return err
+		}
+		cmd.kmsKeyResourceID = strings.TrimSpace(kmsKey)
+	}
```

Comment:
Fair point. The reason is that I left it this way, is that otherwise, it would get fairly complex while I suspect there is no real use-case for this.

The reason for it getting more complex, is that, to make it feel right, you'd have to extract the project id from the given flag and use that for the choice.

At the same time, if you have access to the exact value of 1 of 2 parameters (because you provided is as a flag), you probably also have access to the other one. So copying it should not really be a problem.

In conclusion, I'd go with the simpler solution that covers practically all use-cases and therefore leave it unchanged. What do you say?

Reply:
Ah I see. So the main difference between the two scenarios is that if neither parameters is supplied, we find it okay to prompt for the project ID, while if either of the parameters is supplied, we don't. Seems like a good reason to handle them separately :+1:

About extracting the project ID from the given flag: I agree we can leave this out for now. If it turns out supplying one value is used a lot, we can always add the options to the prompt later.